### PR TITLE
SHIPA-1924: Adding admin email/password authentication

### PR DESCRIPTION
**Pull request by brendonthiede**
This change should not change behavior from 0.0.3 unless an admin_email is provided

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -78,6 +80,18 @@ func WithHost(host string) Option {
 	}
 }
 
+func WithAdminAuth(adminEmail string, adminPassword string) Option {
+	return func(client *Client) error {
+		if adminEmail == "" || adminPassword == "" {
+			return errors.New("admin_email and admin_password can not be empty")
+		}
```
Comment (koncar) [inline client/client.go:87]: we can move this check to `c.authenticate`

File: shipa/provider.go [hunk 2]
```diff
@@ -52,15 +64,26 @@ func Provider() *schema.Provider {
...
-		client.WithToken(token))
+	if adminEmail != "" {
+		c, err = client.NewClient(
+			client.WithHost(host),
+			client.WithAdminAuth(adminEmail, adminPassword))
+	} else {
+		c, err = client.NewClient(
+			client.WithHost(host),
+			client.WithToken(token))
+	}
```
Comment (koncar) [inline shipa/provider.go:86]: i would maybe avoid this if/else

```
c, err = client.NewClient(
			client.WithHost(host),
			client.WithAdminAuth(adminEmail, adminPassword)
                        client.WithToken(token))
```

we can move validation of admnEmail and adminPassword to authenticate, and we trigger authenticate if token is empty, as you do already

File: client/client_test.go [hunk 3]
```diff
@@ -1,19 +1,43 @@
...
+	"testing"
 
 	"github.com/shipa-corp/terraform-provider-shipa/client/clientest"
 )
 
+type AuthRequest struct {
+	Name      string                  `json:"password"`
+	Endpoint  *ClusterEndpointCreate  `json:"endpoint"`
+	Resources *ClusterResourcesCreate `json:"resources,omitempty"`
+}
```
Comment (koncar) [inline client/client_test.go:18]: is this struct used?

File: client/client.go [hunk 4]
```diff
@@ -123,6 +143,29 @@ func (c *Client) doRequest(req *http.Request) ([]byte, int, error) {
 	return body, res.StatusCode, err
 }
 
+func (c *Client) authenticate() error {
+	mapping, encodedData := map[string]string{"password": c.AdminPassword}, new(bytes.Buffer)
+	json.NewEncoder(encodedData).Encode(mapping)
```
Comment (koncar) [inline client/client.go:148]: i think this error is not handled

File: client/client.go [hunk 5]
```diff
@@ -123,6 +143,29 @@ func (c *Client) doRequest(req *http.Request) ([]byte, int, error) {
...
+
+	resp, err := c.HTTPClient.Post(fmt.Sprintf("%s/users/%s/tokens", c.HostURL, c.AdminEmail), "application/json", encodedData)
```
Comment (koncar) [inline client/client.go:150]: maybe introduce function for fmt.Sprintf("%s/users/%s/tokens", c.HostURL, c.AdminEmail)

like 
```
func apiLogin(host, user) {
    return fmt.Sprintf("%s/users/%s/tokens", host, user)
}
```
we are doing it for most of endpoints like:
https://github.com/shipa-corp/terraform-provider-shipa/blob/5fdbd65d2baa35ceda0d83825e61ec2e88259f21/client/client.go#L29-L51

File: client/client.go [hunk 6]
```diff
@@ -123,6 +143,29 @@ func (c *Client) doRequest(req *http.Request) ([]byte, int, error) {
...
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("could not retrieve token for %s", c.AdminEmail)
+	}
+
+	defer resp.Body.Close()
+	var data map[string]interface{}
+	json.NewDecoder(resp.Body).Decode(&data)
```
Comment (koncar) [inline client/client.go:160]: error not handled

**Review by koncar — CHANGES_REQUESTED** [review]

File: shipa/provider.go (hunk 2 above)
Comment (brendonthiede) [reply shipa/provider.go:86]: Currently we validate the token existence up front, whereas authentication check would run for each usage. Delaying the check that email and password are provided also could mean ignoring an empty token up front. What about a "WithAuth" that takes token, email, and password that gives an error of token is empty and both admin email and password are empty? I do still like the check in the authenticate method as well.

**Review by koncar — APPROVED** [review]
looks good

**Pull request merged**